src/main/flight/imu.c

```diff
@@ -261,150 +243,158 @@ static void imuCheckAndResetOrientationQuaternion(const float ax, const float ay
-        ax *= accRecipNorm;
-        ay *= accRecipNorm;
-        az *= accRecipNorm;
+        // Calculate estimated gravity vector in body frame
+        rotateVectorByQuaternion(&vEstGravity, &vGravity, &orientation);    // EF -> BF
 
         // Error is sum of cross product between estimated direction and measured direction of gravity
-        ex = (ay * rMat[2][2] - az * rMat[2][1]);
-        ey = (az * rMat[2][0] - ax * rMat[2][2]);
-        ez = (ax * rMat[2][1] - ay * rMat[2][0]);
+        vectorNormalize(&vAcc, accBF);
+        vectorCrossProduct(&vErr, &vAcc, &vEstGravity);
 
         // Compute and apply integral feedback if enabled
         if (imuRuntimeConfig.dcm_ki_acc > 0.0f) {
             // Stop integrating if spinning beyond the certain limit
             if (spin_rate_sq < sq(DEGREES_TO_RADIANS(SPIN_RATE_LIMIT))) {
-                integralAccX += imuRuntimeConfig.dcm_ki_acc * ex * dt;    // integral error scaled by Ki
-                integralAccY += imuRuntimeConfig.dcm_ki_acc * ey * dt;
-                integralAccZ += imuRuntimeConfig.dcm_ki_acc * ez * dt;
+                t_fp_vector vTmp;
 
-                gx += integralAccX;
-                gy += integralAccY;
-                gz += integralAccZ;
+                // integral error scaled by Ki
+                vectorScale(&vTmp, &vErr, imuRuntimeConfig.dcm_ki_acc * dt);
+                vectorAdd(&vIntegralAcc, &vIntegralAcc, &vTmp);
+
+                // apply integral feedback
+                vectorAdd(&vRotation, &vRotation, &vIntegralAcc);
             }
         }
 
         // Calculate kP gain and apply proportional feedback
-        gx += kpAcc * ex;
-        gy += kpAcc * ey;
-        gz += kpAcc * ez;
+        vectorScale(&vErr, &vErr, imuRuntimeConfig.dcm_kp_acc * imuGetPGainScaleFactor());
+        vectorAdd(&vRotation, &vRotation, &vErr);
     }
 
     // Integrate rate of change of quaternion
-    gx *= (0.5f * dt);
-    gy *= (0.5f * dt);
-    gz *= (0.5f * dt);
-
-    const float qa = q0;
-    const float qb = q1;
-    const float qc = q2;
-    q0 += (-qb * gx - qc * gy - q3 * gz);
-    q1 += (qa * gx + qc * gz - q3 * gy);
-    q2 += (qa * gy - qb * gz + q3 * gx);
-    q3 += (qa * gz + qb * gy - qc * gx);
-
-    // Normalise quaternion
-    const float quatRecipNorm = invSqrt(q0 * q0 + q1 * q1 + q2 * q2 + q3 * q3);
-    q0 *= quatRecipNorm;
-    q1 *= quatRecipNorm;
-    q2 *= quatRecipNorm;
-    q3 *= quatRecipNorm;
+    t_fp_vector vTheta;
+    fpQuaternion_t deltaQ;
+
+    vectorScale(&vTheta, &vRotation, 0.5f * dt);
+    quaternionInitFromVector(&deltaQ, &vTheta);
+    const float thetaMagnitudeSq = vectorNormSquared(&vTheta);
+
+    // Calculate quaternion delta:
+    // Theta is a axis/angle rotation. Direction of a vector is axis, magnitude is angle/2.
+    // Proper quaternion from axis/angle involves computing sin/cos, but the formula becomes numerically unstable as Theta approaches zero.
```

Comment:
The instability is strange - even `sin_approx` should be reasonably accurate for sin(x)/x
(BF approximation polynomials were designed with this in mind)

Reply:
The problem is `/x` part. If our `sin(x)` was perfect it wouldn't matter, but since our sin(x) is an estimate, with `x -> 0` result of `sin_approx(x)/x` diverges from actual value creating a problem.

Reply:
Tried it, `sin_approx(x)/x` correctly converges to 1 and is within 1e-6 absolute error ...
(skipped normalization, valid from -pi to pi )
``` octave
function y = sin_approx(x)
  if(1)
    sinPolyCoef3 =-1.666568107e-1;
    sinPolyCoef5 = 8.312366210e-3;
    sinPolyCoef7 = -1.849218155e-4;
    sinPolyCoef9 = 0;
  else
    sinPolyCoef3 = -1.666665710e-1;                                          % Double: -1.666665709650470145824129400050267289858e-1
    sinPolyCoef5 = 8.333017292e-3;                                          % Double:  8.333017291562218127986291618761571373087e-3
    sinPolyCoef7 = -1.980661520e-4;                                          % Double: -1.980661520135080504411629636078917643846e-4
    sinPolyCoef9 = 2.600054768e-6;                                          % Double:  2.600054767890361277123254766503271638682e-6
  endif

  x(x > .5*pi) = pi - x(x > .5*pi);
  x(x < -.5*pi) = -pi - x(x < -.5*pi);

  x2 = x .* x;
  y= x + x .* x2 .* (sinPolyCoef3 + x2 .* (sinPolyCoef5 + x2 .* (sinPolyCoef7 + x2 .* sinPolyCoef9)));

Reply:
I prefer to be on a safe side here. I would still need to handle a case of no motion (`gyroBF == 0`)

Reply:
It's not about update rate, it's about delta rotation magnitude. At 300Hz IMU if quad is rotating 720deg/s (2.4deg per update). Linear approximation of `sin(x)` has only 0.03% error at this rotation. After 300 iterations (1 second) it will yield 0.8 deg error.

At 100Hz IMU and 720 deg/s rotation error after 1s would be 2 deg already.

In reality there are also measurement errors, delta time jitter etc. All this adds up to rapidly growing error at slow update rates. If we can take out at least interpolation error - I'd say it's worth the effort

Reply:
I'm (slowly) looking into this ... IMO balance is necessary - there is no point using math many orders more accurate than input data / drift ...
At least it's useful to have both available ...